File: crates/server/src/ws/desktop.rs
use axum::extract::ws::{Message, WebSocket};
use futures_util::{SinkExt, StreamExt};
use remote_term_shared::RealtimeMessage;
use std::sync::Arc;
use tokio::sync::mpsc;
use tracing::error;
use uuid::Uuid;

use crate::db::{
    ensure_device_owner, insert_activity_log, mark_device_online, upsert_ai_sessions,
    upsert_projects, upsert_provider_statuses, upsert_sessions, ActivityLogInsert,
};
use crate::state::{AppState, DesktopConnection};
use crate::ws::dispatch::{notify_mobiles, now_heartbeat};

pub async fn desktop_socket(state: Arc<AppState>, user_id: Uuid, socket: WebSocket) {
    let (mut sender, mut receiver) = socket.split();
    let (tx, mut rx) = mpsc::unbounded_channel::<RealtimeMessage>();
    let mut connected_device_id: Option<Uuid> = None;

    let outgoing = tokio::spawn(async move {
        while let Some(message) = rx.recv().await {
            if let Ok(payload) = serde_json::to_string(&message) {
                if sender.send(Message::Text(payload.into())).await.is_err() {
                    break;
                }
            }
        }
    });

    while let Some(Ok(message)) = receiver.next().await {
        if let Message::Text(text) = message {
            match serde_json::from_str::<RealtimeMessage>(&text) {
                Ok(message) => {
                    if let Some(device_id) =
                        handle_desktop_message(&state, user_id, tx.clone(), message).await
                    {
                        connected_device_id = Some(device_id);
                    }
                }
                Err(err) => error!(?err, "invalid desktop websocket payload"),
            }
        }
    }

    if let Some(device_id) = connected_device_id {
        state.desktops.write().await.remove(&device_id);
        if let Err(err) = mark_device_online(&state.pool, device_id, false).await {
            error!(?err, "failed to mark desktop offline");
        }
        insert_activity_log(
            &state.pool,
            ActivityLogInsert {
                user_id,
                device_id: Some(device_id),
                session_id: None,
                kind: "connection",
                title: "桌面代理已离线",
                body: "桌面端 WebSocket 已断开，设备标记为离线。",
                risky: false,
            },
        )
        .await;
        notify_mobiles(&state, user_id, now_heartbeat(device_id)).await;
    }
    outgoing.abort();
}

async fn handle_desktop_message(
    state: &Arc<AppState>,
    user_id: Uuid,
    tx: mpsc::UnboundedSender<RealtimeMessage>,
    message: RealtimeMessage,
) -> Option<Uuid> {
    match message {
        RealtimeMessage::DesktopHeartbeat {
            device_id,
            timestamp: _,
        } => Some(handle_heartbeat(state, user_id, device_id, tx).await),
        RealtimeMessage::SessionsSnapshot {
            device_id,
            sessions,
        } => handle_sessions_snapshot(state, user_id, device_id, sessions).await,
        RealtimeMessage::TerminalOutput { device_id, .. } => {
            notify_mobiles(state, user_id, message).await;
            Some(device_id)
        }
        RealtimeMessage::TerminalError {
            device_id,
            session_id,
            code,
            message: err_message,
        } => {
            handle_terminal_error(
                state,
                user_id,
                device_id,
                session_id.as_deref(),
                &err_message,
            )
            .await;
            notify_mobiles(
                state,
                user_id,
                RealtimeMessage::TerminalError {
                    device_id,
                    session_id,
                    code,
                    message: err_message,
                },
            )
            .await;
            Some(device_id)
        }
        RealtimeMessage::ProvidersSnapshot {
            device_id,
            providers,
        } => handle_providers_snapshot(state, user_id, device_id, providers).await,
        RealtimeMessage::ProjectsSnapshot {
            device_id,
            projects,
        } => handle_projects_snapshot(state, user_id, device_id, projects).await,
        RealtimeMessage::AiSessionsSnapshot {
            device_id,
            sessions,
        } => handle_ai_sessions_snapshot(state, user_id, device_id, sessions).await,
        RealtimeMessage::AiMessageDelta { device_id, .. }
        | RealtimeMessage::AiMessageDone { device_id, .. }
        | RealtimeMessage::AiHistoryResponse { device_id, .. }
        | RealtimeMessage::AiChatOutput { device_id, .. }
        | RealtimeMessage::GitStatusSnapshot {
            snapshot: remote_term_shared::GitStatusSnapshot { device_id, .. },
        } => {
            notify_mobiles(state, user_id, message).await;
            Some(device_id)
        }
        _ => None,
    }
}

async fn handle_heartbeat(
    state: &Arc<AppState>,
    user_id: Uuid,
    device_id: Uuid,
    tx: mpsc::UnboundedSender<RealtimeMessage>,
) -> Uuid {
    if ensure_device_owner(&state.pool, user_id, device_id)
        .await
        .is_err()
    {
        return device_id;
    }
    state
        .desktops
        .write()
        .await
        .insert(device_id, DesktopConnection { user_id, tx });
    if let Err(err) = mark_device_online(&state.pool, device_id, true).await {
        error!(?err, "failed to mark desktop online");
    }
    insert_activity_log(
        &state.pool,
        ActivityLogInsert {
            user_id,
            device_id: Some(device_id),
            session_id: None,
            kind: "connection",
            title: "桌面代理已连接",
            body: "桌面端 WebSocket 心跳已恢复，设备标记为在线。",
            risky: false,
        },
    )
    .await;
    notify_mobiles(state, user_id, now_heartbeat(device_id)).await;
    device_id
}

async fn handle_sessions_snapshot(
    state: &Arc<AppState>,
    user_id: Uuid,
    device_id: Uuid,
    sessions: Vec<remote_term_shared::TerminalSession>,
) -> Option<Uuid> {
    if ensure_device_owner(&state.pool, user_id, device_id)
        .await
        .is_err()
    {
        return None;
    }
    if let Err(err) = upsert_sessions(&state.pool, device_id, &sessions).await {
        error!(?err, "failed to upsert sessions");
    }
    insert_activity_log(
        &state.pool,
        ActivityLogInsert {
            user_id,
            device_id: Some(device_id),
            session_id: None,
            kind: "connection",
            title: "会话快照已同步",
            body: "桌面端已上报最新 tmux/screen 会话列表。",
            risky: false,
        },
    )
    .await;
    notify_mobiles(
        state,
        user_id,
        RealtimeMessage::SessionsSnapshot {
            device_id,
            sessions,
        },
    )
    .await;
    Some(device_id)
}

async fn handle_providers_snapshot(
    state: &Arc<AppState>,
    user_id: Uuid,
    device_id: Uuid,
    providers: Vec<remote_term_shared::DesktopProviderStatus>,
) -> Option<Uuid> {
    if ensure_device_owner(&state.pool, user_id, device_id)
        .await
        .is_err()
    {
        return None;
    }
    if let Err(err) = upsert_provider_statuses(&state.pool, device_id, &providers).await {
        error!(?err, "failed to upsert provider statuses");
    }
    notify_mobiles(
        state,
        user_id,
        RealtimeMessage::ProvidersSnapshot {
            device_id,
            providers,
        },
    )
    .await;
    Some(device_id)
}

async fn handle_projects_snapshot(
    state: &Arc<AppState>,
    user_id: Uuid,
    device_id: Uuid,
    projects: Vec<remote_term_shared::WorkspaceProject>,
) -> Option<Uuid> {
    if ensure_device_owner(&state.pool, user_id, device_id)
        .await
        .is_err()
    {
        return None;
    }
    if let Err(err) = upsert_projects(&state.pool, device_id, &projects).await {
        error!(?err, "failed to upsert projects");
    }
    notify_mobiles(
        state,
        user_id,
        RealtimeMessage::ProjectsSnapshot {
            device_id,
            projects,
        },
    )
    .await;
    Some(device_id)
}

async fn handle_ai_sessions_snapshot(
    state: &Arc<AppState>,
    user_id: Uuid,
    device_id: Uuid,
    sessions: Vec<remote_term_shared::AiSession>,
) -> Option<Uuid> {
    if ensure_device_owner(&state.pool, user_id, device_id)
        .await
        .is_err()
    {
        return None;
    }
    if let Err(err) = upsert_ai_sessions(&state.pool, user_id, device_id, &sessions).await {
        error!(?err, "failed to upsert ai sessions");
    }
    notify_mobiles(
        state,
        user_id,
        RealtimeMessage::AiSessionsSnapshot {
            device_id,
            sessions,
        },
    )
    .await;
    Some(device_id)
}

async fn handle_terminal_error(
    state: &Arc<AppState>,
    user_id: Uuid,
    device_id: Uuid,
    session_id: Option<&str>,
    err_message: &str,
) {
    insert_activity_log(
        &state.pool,
        ActivityLogInsert {
            user_id,
            device_id: Some(device_id),
            session_id,
            kind: "error",
            title: "终端错误",
            body: err_message,
            risky: false,
        },
    )
    .await;
}

File: backend/internal/ws/desktop.go
// Desktop-side WebSocket message handling. Ports crates/server/src/ws/desktop.rs.
//
// DB side effects mirror the Rust source exactly:
//   - DesktopHeartbeat: ensure_device_owner + mark_device_online(true) +
//     insert_activity_log("桌面代理已连接") + notify_mobiles(heartbeat).
//   - ProvidersSnapshot: ensure_device_owner + upsert_provider_statuses +
//     notify_mobiles(message).
//   - ProjectsSnapshot: ensure_device_owner + upsert_projects +
//     notify_mobiles(message).
//   - AiSessionsSnapshot: ensure_device_owner + upsert_ai_sessions +
//     notify_mobiles(message).
//   - AiMessageDelta / AiMessageDone / AiHistoryResponse / AiChatOutput /
//     GitStatusSnapshot: pure forward to mobiles, no DB writes.
//
// The desktop is registered in AppState at upgrade time (see ws.go), so
// handleDesktopHeartbeat does not re-register — unlike Rust, which lazily
// registers on the first heartbeat. This is the one deliberate behavioral
// difference required by the Go state.Connection API (AddDesktop starts the
// write goroutine and must be called before any outbound send).
package ws

import (
	"context"
	"log"

	"github.com/google/uuid"

	"github.com/gaolin89898/ai-workbench/backend/internal/models"
	"github.com/gaolin89898/ai-workbench/backend/internal/protocol"
)

// handleDesktopMessage dispatches an inbound desktop message. It returns true
// when the message type was recognized (matching the Rust Option::Some return
// that sets connected_device_id), so readLoop can track whether disconnect
// cleanup should emit DB side effects.
func (h *Handler) handleDesktopMessage(msg protocol.Message, userID, deviceID uuid.UUID) bool {
	switch m := msg.(type) {
	case protocol.DesktopHeartbeat:
		h.handleDesktopHeartbeat(userID, m)
		return true
	case protocol.ProvidersSnapshot:
		h.handleProvidersSnapshot(userID, m)
		return true
	case protocol.ProjectsSnapshot:
		h.handleProjectsSnapshot(userID, m)
		return true
	case protocol.AiSessionsSnapshot:
		h.handleAiSessionsSnapshot(userID, m)
		return true
	case protocol.AiMessageDelta, protocol.AiMessageDone, protocol.AiHistoryResponse, protocol.AiChatOutput, protocol.GitStatusSnapshotMessage:
		// Pure forward to mobiles; no DB side effects.
		h.notifyMobiles(userID, msg)
		return true
	default:
		// Unknown/unhandled desktop message: ignore (matches Rust `_ => None`).
		return false
	}
}

// handleDesktopHeartbeat mirrors handle_heartbeat in desktop.rs: verify device
// ownership, mark the device online, log the connection, and notify mobiles
// with a fresh heartbeat. The Rust version inserts the desktop into
// state.desktops here; in the Go port the connection is already registered at
// upgrade time, so that step is skipped.
func (h *Handler) handleDesktopHeartbeat(userID uuid.UUID, m protocol.DesktopHeartbeat) {
	if m.DeviceId == "" {
		return
	}
	ctx := context.Background()
	if err := h.DB.EnsureDeviceOwner(ctx, userID.String(), m.DeviceId); err != nil {
		return
	}
	if err := h.DB.MarkDeviceOnline(ctx, m.DeviceId, true); err != nil {
		log.Printf("ws: failed to mark desktop online: %v", err)
	}
	devID := m.DeviceId
	if err := h.DB.InsertActivityLog(ctx, models.ActivityLogInsert{
		UserId:   userID.String(),
		DeviceId: &devID,
		Kind:     "connection",
		Title:    "桌面代理已连接",
		Body:     "桌面端 WebSocket 心跳已恢复，设备标记为在线。",
		Risky:    false,
	}); err != nil {
		log.Printf("ws: failed to insert desktop-online activity log: %v", err)
	}
	if devUUID, err := uuid.Parse(m.DeviceId); err == nil {
		h.notifyMobilesHeartbeat(userID, devUUID)
	}
}

// handleProvidersSnapshot mirrors handle_providers_snapshot in desktop.rs.
func (h *Handler) handleProvidersSnapshot(userID uuid.UUID, m protocol.ProvidersSnapshot) {
	if m.DeviceId == "" {
		return
	}
	ctx := context.Background()
	if err := h.DB.EnsureDeviceOwner(ctx, userID.String(), m.DeviceId); err != nil {
		return
	}
	if err := h.DB.UpsertProviderStatuses(ctx, m.DeviceId, toModelProviders(m.Providers)); err != nil {
		log.Printf("ws: failed to upsert provider statuses: %v", err)
	}
	h.notifyMobiles(userID, m)
}

// handleProjectsSnapshot mirrors handle_projects_snapshot in desktop.rs.
func (h *Handler) handleProjectsSnapshot(userID uuid.UUID, m protocol.ProjectsSnapshot) {
	if m.DeviceId == "" {
		return
	}
	ctx := context.Background()
	if err := h.DB.EnsureDeviceOwner(ctx, userID.String(), m.DeviceId); err != nil {
		return
	}
	if err := h.DB.UpsertProjects(ctx, m.DeviceId, toModelProjects(m.Projects)); err != nil {
		log.Printf("ws: failed to upsert projects: %v", err)
	}
	h.notifyMobiles(userID, m)
}

// handleAiSessionsSnapshot mirrors handle_ai_sessions_snapshot in desktop.rs.
// The Rust upsert_ai_sessions takes user_id and device_id; the Go port passes
// the same.
func (h *Handler) handleAiSessionsSnapshot(userID uuid.UUID, m protocol.AiSessionsSnapshot) {
	if m.DeviceId == "" {
		return
	}
	ctx := context.Background()
	if err := h.DB.EnsureDeviceOwner(ctx, userID.String(), m.DeviceId); err != nil {
		return
	}
	if err := h.DB.UpsertAiSessions(ctx, userID.String(), m.DeviceId, toModelAiSessions(m.Sessions)); err != nil {
		log.Printf("ws: failed to upsert ai sessions: %v", err)
	}
	h.notifyMobiles(userID, m)
}

// cleanupDesktop tears down a desktop connection: remove it from AppState,
// mark the device offline, log the disconnection, and notify mobiles with a
// fresh heartbeat. DB side effects only run if the connection was activated
// (at least one inbound message handled), mirroring the Rust
// connected_device_id guard in desktop_socket.
func (h *Handler) cleanupDesktop(userID, deviceID uuid.UUID, activated bool) {
	h.State.RemoveDesktop(deviceID)
	if !activated {
		return
	}
	ctx := context.Background()
	if err := h.DB.MarkDeviceOnline(ctx, deviceID.String(), false); err != nil {
		log.Printf("ws: failed to mark desktop offline: %v", err)
	}
	devID := deviceID.String()
	if err := h.DB.InsertActivityLog(ctx, models.ActivityLogInsert{
		UserId:   userID.String(),
		DeviceId: &devID,
		Kind:     "connection",
		Title:    "桌面代理已离线",
		Body:     "桌面端 WebSocket 已断开，设备标记为离线。",
		Risky:    false,
	}); err != nil {
		log.Printf("ws: failed to insert desktop-offline activity log: %v", err)
	}
	h.notifyMobilesHeartbeat(userID, deviceID)
}

// ---- protocol -> models converters ----
//
// protocol and models define parallel structs (protocol carries the wire
// layout, models the DB column layout). The db upsert helpers take []models.X,
// so inbound protocol payloads are converted here. Nullable pointer fields and
// time.Time fields are copied by value.

func toModelProviders(ps []protocol.DesktopProviderStatus) []models.DesktopProviderStatus {
	out := make([]models.DesktopProviderStatus, len(ps))
	for i, p := range ps {
		out[i] = models.DesktopProviderStatus{
			ProviderId:    p.ProviderId,
			Installed:     p.Installed,
			Version:       p.Version,
			AuthStatus:    string(p.AuthStatus),
			LastCheckedAt: p.LastCheckedAt,
		}
	}
	return out
}

func toModelProjects(ps []protocol.WorkspaceProject) []models.WorkspaceProject {
	out := make([]models.WorkspaceProject, len(ps))
	for i, p := range ps {
		out[i] = models.WorkspaceProject{
			Id:        p.Id,
			DeviceId:  p.DeviceId,
			Name:      p.Name,
			Path:      p.Path,
			GitBranch: p.GitBranch,
			GitDirty:  p.GitDirty,
			UpdatedAt: p.UpdatedAt,
		}
	}
	return out
}

func toModelAiSessions(ss []protocol.AiSession) []models.AiSession {
	out := make([]models.AiSession, len(ss))
	for i, s := range ss {
		out[i] = models.AiSession{
			Id:                s.Id,
			UserId:            s.UserId,
			DeviceId:          s.DeviceId,
			ProjectId:         s.ProjectId,
			ProviderId:        s.ProviderId,
			TerminalSessionId: s.TerminalSessionId,
			ProviderSessionId: s.ProviderSessionId,
			Title:             s.Title,
			Status:            string(s.Status),
			Summary:           s.Summary,
			ArchivedAt:        s.ArchivedAt,
			UpdatedAt:         s.UpdatedAt,
		}
	}
	return out
}
